### PR TITLE
[Modules] Updated Compute/VirtualMachineScaleSets to new dependency approach

### PR DESCRIPTION
**Pull request by AlexanderSehr**
# Description

- Updated Compute/VirtualMachineScaleSets to new dependency approach

| Pipeline |
| - |
| [Compute: VirtualMachineScaleSets](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.virtualmachinescalesets.yml) |
| [Compute: VirtualMachines](https://github.com/Azure/ResourceModules/actions/workflows/ms.compute.virtualmachines.yml) |

# Type of Change

Please delete options that are not relevant.

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Update to documentation

### CODE REVIEW

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/deploy.test.bicep [hunk 1]
```diff
@@ -0,0 +1,86 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+@description('Optional. The name of the resource group to deploy for a testing purposes.')
```
Comment (MrMCake) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/deploy.test.bicep:6]:
```suggestion
@description('Optional. The name of the resource group to deploy for testing purposes.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep [hunk 2]
```diff
@@ -0,0 +1,183 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+@description('Optional. The name of the resource group to deploy for a testing purposes.')
```
Comment (MrMCake) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep:6]:
```suggestion
@description('Optional. The name of the resource group to deploy for testing purposes.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows.min/deploy.test.bicep [hunk 3]
```diff
@@ -0,0 +1,81 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+@description('Optional. The name of the resource group to deploy for a testing purposes.')
```
Comment (MrMCake) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows.min/deploy.test.bicep:6]:
```suggestion
@description('Optional. The name of the resource group to deploy for testing purposes.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/deploy.test.bicep [hunk 4]
```diff
@@ -0,0 +1,179 @@
+targetScope = 'subscription'
+
+// ========== //
+// Parameters //
+// ========== //
+@description('Optional. The name of the resource group to deploy for a testing purposes.')
```
Comment (MrMCake) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/deploy.test.bicep:6]:
```suggestion
@description('Optional. The name of the resource group to deploy for testing purposes.')
```

File: utilities/tools/Set-ModuleReadMe.ps1 [hunk 5]
```diff
@@ -627,8 +627,11 @@ Convert the given Bicep parameter block to JSON parameter block
 .PARAMETER BicepParamBlock
 Mandatory. The Bicep parameter block to process
 
+.PARAMETER CurrentFilePath
```
Comment (eriqua) [inline utilities/tools/Set-ModuleReadMe.ps1:630]: Is this change related to the VMSS in particular or can it be integrated into the Set-ModuleReadme PR?

**Review by eriqua — COMMENTED** [review]
The validation pipeline started failing on the API version Pester test, so that should also be updated https://github.com/Azure/ResourceModules/actions/runs/3436153620/jobs/5729347182#step:4:556

File: utilities/tools/Set-ModuleReadMe.ps1 (hunk 5 above)
Comment (AlexanderSehr) [reply utilities/tools/Set-ModuleReadMe.ps1]: This was indeed already in main by now. I merged latest main & resolved the conflict

**Comment by AlexanderSehr** [comment]
> The validation pipeline started failing on the API version Pester test, so that should also be updated https://github.com/Azure/ResourceModules/actions/runs/3436153620/jobs/5729347182#step:4:556

Works now :)

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1 [hunk 6]
```diff
@@ -0,0 +1,19 @@
+﻿param(
+    [string] $SSHKeyName,
+    [string] $ResourceGroupName
+)
+
+if (-not ($sshKey = Get-AzSshKey -ResourceGroupName $ResourceGroupName | Where-Object { $_.Name -eq $SSHKeyName })) {
+    Write-Verbose "No SSH key [$SSHKeyName] found in Resource Group [$ResourceGroupName]. Generating new." -Verbose
+    $null = ssh-keygen -f generated -N (Get-Random -Maximum 99999)
+    $publicKey = Get-Content 'generated.pub' -Raw
+    # $privateKey = cat generated | Out-String
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1:10]: Should this be removed?

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    }
+}
+
+resource managedIdentity 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
+    name: managedIdentityName
+    location: location
+}
+
+resource msiRGContrRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+    name: guid('msi-${resourceGroup().id}-${location}-${managedIdentity.id}-RG-Reader-RoleAssignment')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:42]: Align with similar resources in other modules dependencies e.g.
```suggestion
    name: guid(resourceGroup().id, 'Contributor', managedIdentity.id)
```
In any case replace "Reader" with "Contributor

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    location: location
+    kind: 'AzurePowerShell'
+    identity: {
+        type: 'UserAssigned'
+        userAssignedIdentities: {
+            '${managedIdentity.id}': {}
+        }
+    }
+    properties: {
+        azPowerShellVersion: '6.2.1'
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:62]: I'd suggest to use a more recent version

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    }
+}
+
+@description('The resource ID of the created Virtual Network Subnet.')
+output subnetResourceId string = virtualNetwork.properties.subnets[0].id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created SSH Key')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:86]:
```suggestion
@description('The resource ID of the created SSH Key.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+output SSHKeyResourceID string = sshKey.id
+
+@description('The Public Key of the created SSH Key')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:89]:
```suggestion
@description('The Public Key of the created SSH Key.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep [hunk 11]
```diff
@@ -0,0 +1,196 @@
+@description('Optional. The location to deploy to')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep:1]:
```suggestion
@description('Optional. The location to deploy to.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 12]
```diff
@@ -0,0 +1,90 @@
+@description('Optional. The location to deploy to')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:1]:
```suggestion
@description('Optional. The location to deploy to.')
```

Probably in other modules as well

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1 (hunk 6 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1:10]: I'd suggest to keep it. It essentially shows you how you 'could' also retrieve the private key from what the `ssh-keygen` function produced. It just happends that we don't need it in our test - hence it is only a comment. However, I don't know about you, but I often re-use these kinds of scripts outside of their original context ;) Hence, I guess no harm in keeping it,

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 7 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: Updated - also for VM if it's consistency we're after

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 8 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:62]: Fair enough - thow I gotta ask if that's really necessary as it works and is only a test

Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: Updated to 9.1

File: modules/Microsoft.Compute/virtualMachines/.test/linux.atmg/dependencies.bicep [hunk 13]
```diff
@@ -59,7 +59,7 @@ resource sshDeploymentScript 'Microsoft.Resources/deploymentScripts@2020-10-01'
     }
   }
   properties: {
-    azPowerShellVersion: '6.2.1'
+    azPowerShellVersion: '9.1'
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachines/.test/linux.atmg/dependencies.bicep:62]: Missing update to msiRGContrRoleAssignment name above as done for others

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1 (hunk 6 above)
Comment (eriqua) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/.scripts/New-SSHKey.ps1:10]: Fair enough :)

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+output subnetResourceId string = virtualNetwork.properties.subnets[0].id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created SSH Key.')
+output SSHKeyResourceID string = sshKey.id
+
+@description('The Public Key of the created SSH Key.')
+output SSHKey string = sshKey.properties.publicKey
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:90]: I'd suggest to change all occurrences
```suggestion
output SSHKeyPublicKey string = sshKey.properties.publicKey
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+@description('The resource ID of the created Virtual Network Subnet.')
+output subnetResourceId string = virtualNetwork.properties.subnets[0].id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
+
+@description('The resource ID of the created SSH Key.')
+output SSHKeyResourceID string = sshKey.id
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:87]: If not used I'd suggest to remove the output

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+    properties: {
+        publicKey: sshDeploymentScript.properties.outputs.publicKey
+    }
+}
+
+@description('The resource ID of the created Virtual Network Subnet.')
+output subnetResourceId string = virtualNetwork.properties.subnets[0].id
+
+@description('The principal ID of the created Managed Identity.')
+output managedIdentityPrincipalId string = managedIdentity.properties.principalId
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:84]: If not used I'd suggest to remove the output

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+output storageAccountName string = storageAccount.name
+
+@description('The resource ID of the created Storage Account.')
+output storageAccountResourceId string = storageAccount.id
+
+@description('The URL of the Custom Script Extension in the created Storage Account')
+output storageAccountCSEFileUrl string = '${storageAccount.properties.primaryEndpoints.blob}${storageContainerName}/${storageAccountCSEFileName}'
+
+@description('The resource ID of the created SSH Key')
+output SSHKeyResourceID string = sshKey.id
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep:193]: If not used I'd suggest to remove the output

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+      }
+    }
+    extensionDependencyAgentConfig: {
+      enabled: true
+    }
+    extensionDiskEncryptionConfig: {
+      enabled: true
+      settings: {
+        EncryptionOperation: 'EnableEncryption'
+        KekVaultResourceId: resourceGroupResources.outputs.keyVaultResourceId
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep:128]: Looks duplicated

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows.min/dependencies.bicep [hunk 19]
```diff
@@ -0,0 +1,28 @@
+@description('Optional. The location to deploy to')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows.min/dependencies.bicep:1]:
```suggestion
@description('Optional. The location to deploy to.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/dependencies.bicep [hunk 20]
```diff
@@ -0,0 +1,166 @@
+@description('Optional. The location to deploy to')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/dependencies.bicep:1]:
```suggestion
@description('Optional. The location to deploy to.')
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/dependencies.bicep [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+    scope: resourceGroup()
+    properties: {
+        principalId: managedIdentity.properties.principalId
+        roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c') // Contributor
+        principalType: 'ServicePrincipal'
+    }
+}
+
+resource msiKVCryptoUserRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+    name: guid('msi-${keyVault::key.id}-${location}-${managedIdentity.id}-KeyVault-Key-Read-RoleAssignment')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/windows/dependencies.bicep:86]: This should probably also use the same approach
```suggestion
    name: guid(keyVault::key.id, 'Key Vault Crypto User', managedIdentity.id)
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+    scope: resourceGroup()
+    properties: {
+        principalId: managedIdentity.properties.principalId
+        roleDefinitionId: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'b24988ac-6180-42a0-ab88-20f7382dd24c') // Contributor
+        principalType: 'ServicePrincipal'
+    }
+}
+
+resource msiKVCryptoUserRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
+    name: guid('msi-${keyVault::key.id}-${location}-${managedIdentity.id}-KeyVault-Key-Read-RoleAssignment')
```
Comment (eriqua) [inline modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep:89]:
```suggestion
    name: guid(keyVault::key.id, 'Key Vault Crypto User', managedIdentity.id)
```

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep (hunk 18 above)
Comment (eriqua) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/deploy.test.bicep:128]: my bad reading, it's not

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 8 above)
Comment (eriqua) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: Probably too recent https://github.com/Azure/ResourceModules/actions/runs/3474887217/jobs/5808567494#step:4:602

I guess 8.0 or 9.0 would be fine

Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: updated to 9.0 and will re-test once the other comments are addressed

File: modules/Microsoft.Compute/virtualMachines/.test/linux.atmg/dependencies.bicep (hunk 13 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachines/.test/linux.atmg/dependencies.bicep]: done

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 14 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: done

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 15 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep:87]: done

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep (hunk 16 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux.min/dependencies.bicep]: done

File: modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep (hunk 17 above)
Comment (AlexanderSehr) [reply modules/Microsoft.Compute/virtualMachineScaleSets/.test/linux/dependencies.bicep]: done

**Review by eriqua — APPROVED** [review]
🆗

**Pull request merged**